flyteplugins/go/tasks/pluginmachinery/ioutils/remote_file_output_reader.go

```diff
@@ -3,75 +3,243 @@ package ioutils
 			Message: errorDoc.Error.Message,
 			Kind:    errorDoc.Error.Origin,
+			Worker:  errorDoc.Error.Worker,
 		},
 	}
 
 	if errorDoc.Error.Kind == core.ContainerError_RECOVERABLE {
-		ee.IsRecoverable = true
+		executionError.IsRecoverable = true
+	}
+
+	return executionError
+}
+
+func (s *SingleFileErrorRetriever) GetError(ctx context.Context) (io.ExecutionError, error) {
+	errorDoc := &core.ErrorDocument{}
+	err := s.store.ReadProtobuf(ctx, s.errorFilePath, errorDoc)
+	if err != nil {
+		if storage.IsNotFound(err) {
+			return io.ExecutionError{
+				IsRecoverable: true,
+				ExecutionError: &core.ExecutionError{
+					Code:    "ErrorFileNotFound",
+					Message: err.Error(),
+					Kind:    core.ExecutionError_SYSTEM,
+				},
+			}, nil
+		}
+		return io.ExecutionError{}, errors.Wrapf(err, "failed to read error data from task @[%s]", s.errorFilePath)
+	}
+
+	return errorDoc2ExecutionError(errorDoc, s.errorFilePath), nil
+}
+
+type EarliestFileErrorRetriever struct {
+	baseErrorRetriever
+	errorDirPath           storage.DataReference
+	canonicalErrorFilename string
+}
+
+func (e *EarliestFileErrorRetriever) parseErrorFilename() (errorFilePathPrefix storage.DataReference, errorFileExtension string, err error) {
+	// If the canonical error file name is error.pb, we expect multiple error files
+	// to have name error<suffix>.pb
+	pieces := strings.Split(e.canonicalErrorFilename, ".")
+	if len(pieces) != 2 {
+		err = errors.Errorf("expected canonical error filename to have a single dot (.), got %d", len(pieces))
+		return
+	}
+	errorFilePrefix := pieces[0]
+	scheme, container, key, _ := e.errorDirPath.Split()
+	errorFilePathPrefix = storage.NewDataReference(scheme, container, filepath.Join(key, errorFilePrefix))
+	errorFileExtension = fmt.Sprintf(".%s", pieces[1])
+	return
+}
+
+func (e *EarliestFileErrorRetriever) HasError(ctx context.Context) (bool, error) {
+	errorFilePathPrefix, errorFileExtension, err := e.parseErrorFilename()
+	if err != nil {
+		return false, errors.Wrapf(err, "failed to parse canonical error filename @[%s]", e.canonicalErrorFilename)
+	}
+	hasError := false
+	const maxItems = 1000
+	cursor := storage.NewCursorAtStart()
+	for cursor != storage.NewCursorAtEnd() {
+		var err error
+		var errorFilePaths []storage.DataReference
+		errorFilePaths, cursor, err = e.store.List(ctx, errorFilePathPrefix, maxItems, cursor)
+		if err != nil {
+			return false, errors.Wrapf(err, "failed to list error files @[%s]", e.errorDirPath)
+		}
+		for _, errorFilePath := range errorFilePaths {
+			if strings.HasSuffix(errorFilePath.String(), errorFileExtension) {
+				metadata, err := e.store.Head(ctx, errorFilePath)
+				if err != nil {
+					return false, errors.Wrapf(err, "failed to read error file @[%s]", errorFilePath)
+				}
+				err = e.validatePayloadSize(metadata)
+				if err != nil {
+					return false, errors.Wrapf(err, "error file @[%s] failed payload size validation", errorFilePath)
+				}
+				hasError = true
```

Comment:
Shouldn't we return immediately here as soon as we found at least one error?

Reply:
That was my first instinct, but wanted to run validatePayloadSize for all files to keep it consistent with the existing logic.